### ISSUE TITLE
DTLS v1.2  connect and read with SSL_ERROR_WANT_READ

### ISSUE THREAD

**Issue by sasakiakaya** [openssl/openssl#24022]
HI,

SSL_connect and SSL_read has very high error frequency with SSL_ERROR_WANT_READ. This led to the failure of SSL connection and reception.
I‘ve dump the UDP package from socket, and I have ensured that the socket receives complete packets.

Also, the handshake processs is weird. It seems that SSL do handshake twice.

tcpdump:
![tcpdump](https://github.com/openssl/openssl/assets/45138606/7f4711a3-e98a-49f5-972c-4e9013755bc6)

OS: openwrt 5.15
Arch: aarch64
openssl ver: 3.0.10

**Comment by nhorman** [comment]
Can you upload the tcpdump in pcap format?  Its impossible to tell from the above, but it initially looks like you have two clients connecting from the same source ip/port value

**Comment by sasakiakaya** [comment]
> Can you upload the tcpdump in pcap format? Its impossible to tell from the above, but it initially looks like you have two clients connecting from the same source ip/port value

[tcpdump.zip](https://github.com/openssl/openssl/files/14833887/tcpdump.zip)

I have checked my code and "netstat", I'm sure there is only one UDP socket fd for this ID, and SSL_connect is only called once

**Comment by sasakiakaya** [comment]
SSL_connect has very high error frequency with SSL_ERROR_WANT_READ,  and the number of errors is random. When SSL_connect  return -1 and get error code = SSL_ERROR_WANT_READ, I try to call SSL_connect again. 
What’s certain is that when the SSL handshake is successful, all data will be received twice, and one of them will lead to SSL_ERROR_WANT_READ.

**Comment by nhorman** [comment]
Ah, that might be the issue.  From https://www.openssl.org/docs/man1.0.2/man3/SSL_connect.html :
```
If the underlying BIO is non-blocking, SSL_connect() will also return when the underlying BIO could not satisfy the needs of SSL_connect() to continue the handshake, indicating the problem by the return value -1. In this case a call to SSL_get_error() with the return value of SSL_connect() will yield SSL_ERROR_WANT_READ or SSL_ERROR_WANT_WRITE. The calling process then must repeat the call after taking appropriate action to satisfy the needs of SSL_connect(). The action depends on the underlying BIO. When using a non-blocking socket, nothing is to be done, but select() can be used to check for the required condition. When using a buffering BIO, like a BIO pair, data must be written into or retrieved out of the BIO before being able to continue.
```

If you are using a non-blocking socket, you shouldn't call SSL_connect again, but instead use select to wait for the required condition

**Comment by sasakiakaya** [comment]
> Ah, that might be the issue. From https://www.openssl.org/docs/man1.0.2/man3/SSL_connect.html :
> 
> ```
> If the underlying BIO is non-blocking, SSL_connect() will also return when the underlying BIO could not satisfy the needs of SSL_connect() to continue the handshake, indicating the problem by the return value -1. In this case a call to SSL_get_error() with the return value of SSL_connect() will yield SSL_ERROR_WANT_READ or SSL_ERROR_WANT_WRITE. The calling process then must repeat the call after taking appropriate action to satisfy the needs of SSL_connect(). The action depends on the underlying BIO. When using a non-blocking socket, nothing is to be done, but select() can be used to check for the required condition. When using a buffering BIO, like a BIO pair, data must be written into or retrieved out of the BIO before being able to continue.
> ```
> 
> If you are using a non-blocking socket, you shouldn't call SSL_connect again, but instead use select to wait for the required condition

But DTLS is over UDP, and UDP is blocking socket.

**Comment by nhorman** [comment]
The choice to use UDP is independent of the choice to use a blocking socket

**Comment by sasakiakaya** [comment]
> The choice to use UDP is independent of the choice to use a blocking socket. What you're tcpdump is suggesting is that you are using a nonblocking UDP socket (e.g a BIO_dgram object with BIO_socket_nbio called on it with BIO_SOCK_NONBLOCK set

If the socket is non-blocking, I think the package from socket will be truncated. But I've tried to dump the UDP package by recvfrom(), the package is intact.

**Comment by mattcaswell** [comment]
> If the socket is non-blocking, I think the package from socket will be truncated. 

Why? It is common to run UDP with a non-blocking socket.

**Comment by sasakiakaya** [comment]
> > If the socket is non-blocking, I think the package from socket will be truncated.
> 
> Why? It is common to run UDP with a non-blocking socket.

Sorry. I misunderstood the meaning of "non-blocking". My code is as below:

numBytes = recvfrom(fd, buffer, 1024,MSG_PEEK, (struct sockaddr*)&addr_str, (int32_t*)&addrLen);
if(numBytes  > 0)
  SSL_read(ssl, buffer, 1024);

If socket is non-blocking, recvfrom() may return with EWOULDBLOCK. I've check my log, recvfrom() fetches the full UDP package each time it is called

**Comment by nhorman** [comment]
No, it won't, at least not universally.  Only if there is no data on the receive queue will it return EWOULDBLOCK (or more specifically EAGAIN I believe).  If there is any data available to receive it will return that data and indicate the number of bytes filled in your provided buffer.  With UDP, just because you request X bytes, doesn't mean you will receive X bytes, you may receive less, and that is not a blocking condition.

You shouldn't use the above method as a check to see if SSL_read is callable.  You should instead call SSL_read and, if an error is reported, check the SSL error condition for SSL_ERROR_WANT_READ or SSL_ERROR_WANT_write to preform a subsequent select or poll operation on the connection ( you can see examples in the doc/designs/ddd subdirectory of the project)

The same holds true with the SSL_connect call.

**Comment by kroeckx** [comment]
UDP preserves message boundaries, you always read 1 whole packet from the kernel. It will not combine packets. If your buffer is too small, the rest of the packet is discarded. This has nothing to do with blocking on non-blocking.

The difference between blocking and non-blocking mode is that in blocking mode, the recvfrom() will hang until data is available to be returned. In non-blocking mode it will return and set EGAIN or EWOULDBLOCK if there is currently no data. In non-blocking mode you want to wait with something like select or poll to wait until data is available, and not call recvfrom() in a loop without some kind of delay.

Note that on most but not all systems EWOULDBLOCK and EGAIN have the same value, so you should check both values.
